Enabled masked methods for bfloat16

izdeby opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #24224 Enabled torch.mm and torch.mv for bfloat16 on CPU
* **#24183 Enabled masked methods for bfloat16**
-----------
Fix: Enabled masked select/scatter/fill  for BFloat16 on CPU
Test: via unit tests

Differential Revision: [D16763461](https://our.internmc.facebook.com/intern/diff/D16763461)

izdeby committed:
Enabled masked methods for bfloat16

gchanan commented on test/test_torch.py:
```diff
@@ -8409,29 +8409,33 @@ def test_scatter_add_bool(self):
+            for maskType in [torch.uint8, torch.bool]:
+                for dt in torch.testing.get_all_dtypes():
+                    if dt == torch.half or dt == torch.bool:
+                        continue
```
first, do half and double not actually work for masking?

second, as has been mentioned in a number of reviews, don't skip tests, assert they don't work with a message saying to enable once it works.  As you've done it here, we will never actually run these tests, even when we enable bool and half.

izdeby replied:
double?

ezyang replied:
(probably meant bool)

izdeby committed:
Update on "Enabled masked methods for bfloat16"

-----------
Fix: Enabled masked select/scatter/fill  for BFloat16 on CPU
Test: via unit tests

Differential Revision: [D16763461](https://our.internmc.facebook.com/intern/diff/D16763461)

ezyang commented on test/test_torch.py:
```diff
@@ -8454,29 +8454,40 @@ def test_scatter_add_bool(self):
+
+                    if dt == torch.bool:
+                        # torch.bool is a special case and is being tested
+                        # in a separate test
```
...because bool will truncate elements outside of its range? (Spell it out for the reader ;)

gchanan replied:
...but that's not actually all that special :).  I don't know why we just don't have an if statement that sets up the destination for the case where it's bool -- we don't test the edge cases for bool tensors for what seems like non-compelling reasons.

izdeby replied:
I find this a bit more cleaner. in case when there are a lot of if/else conditions, its hard to figure out what does the test do quickly. But if you insist, i will combine all masked_* tests with their masked_*_bool versions

gchanan commented on test/test_torch.py:
```diff
@@ -8454,29 +8454,40 @@ def test_scatter_add_bool(self):
 
     def test_masked_scatter(self):
         with warnings.catch_warnings(record=True) as w:
```
unconditionally catching the warnings and not checking them doesn't seem great -- we don't actually check that we raise warnings for uint8 and if we start raising warnings for bool no one will notice.

izdeby committed:
Update on "Enabled masked methods for bfloat16"

-----------
Fix: Enabled masked select/scatter/fill  for BFloat16 on CPU
Test: via unit tests

Differential Revision: [D16763461](https://our.internmc.facebook.com/intern/diff/D16763461)

izdeby committed:
Update on "Enabled masked methods for bfloat16"

-----------
Fix: Enabled masked select/scatter/fill  for BFloat16 on CPU
Test: via unit tests

Differential Revision: [D16763461](https://our.internmc.facebook.com/intern/diff/D16763461)
